media: ov9282: Add external trigger mode support

omeredemen opened the pull request:
Adds DT property `trigger-mode` to enable FSIN-triggered frame capture. Includes overlay and README update for ov9281_trig.

https://github.com/raspberrypi/linux/issues/5349#issuecomment-3058162605

pelwell commented:
Please squash your ov9282 commits into one, then split that in two - one for the driver changes and one for the overlay changes. You can then run `scripts/checkpatch.pl -g HEAD^` and `scripts/checkpatch.pl -g HEAD` to check for remaining indentation complaints. It would also be nice if the Signed-off-by lines weren't "Your Name". After all that, use `git push -f ...` to update your branch (and hence the PR).

With those changes, I presume you are happy with this @6by9?

6by9 commented on drivers/media/i2c/ov9282.c:
```diff
@@ -992,15 +1060,25 @@ static int ov9282_start_streaming(struct ov9282 *ov9282)
 	/* Start streaming */
 	ret = ov9282_write_reg(ov9282, OV9282_REG_MODE_SELECT,
-			       1, OV9282_MODE_STREAMING);
+			1, OV9282_MODE_STREAMING);
```
Odd whitespace change for no reason

6by9 commented on drivers/media/i2c/ov9282.c:
```diff
@@ -992,15 +1060,25 @@ static int ov9282_start_streaming(struct ov9282 *ov9282)
 
 	/* Setup handler will write actual exposure and gain */
-	ret =  __v4l2_ctrl_handler_setup(ov9282->sd.ctrl_handler);
+	ret = __v4l2_ctrl_handler_setup(ov9282->sd.ctrl_handler);
```
Whitespace change

6by9 replied:
Looking again I'll acknowledge that it's wrong in the original, but changing it should be a separate patch.

6by9 commented on drivers/media/i2c/ov9282.c:
```diff
@@ -947,6 +956,65 @@ static int ov9282_get_selection(struct v4l2_subdev *sd,
+
+	/* 1 frame per trigger */
+        ov9282_write_reg(ov9282, OV9282_REG_NUM_FRAME_ON_TRIG, 1, 0x01);
+        if (ret) {
```
You've not saved the result of `ov9282_write_reg` to `ret`, so this will never fail.

Ditto subsequent calls.

6by9 reviewed (changes requested):
Sorry for the delay in reviewing.

omeredemen commented:
> I need external trigger in order to capture laser light reflected from a moving mirror, so I tried this code with:
> 
> `sudo rpi-update pulls/6954`
> 
> It works. Tested with CAM-MIPIOV9281V2 (Innomaker) on a Raspberry 5. If I can help somehow, please let me know. This should be made available in the official kernel (not everybody can find this pull request)
> 
> Thank you.

Sorry for the latency. @stalinbeltran i am glad it worked for you.

6by9 commented on drivers/media/i2c/ov9282.c:
```diff
@@ -992,15 +1050,24 @@ static int ov9282_start_streaming(struct ov9282 *ov9282)
 	/* Start streaming */
-	ret = ov9282_write_reg(ov9282, OV9282_REG_MODE_SELECT,
-			       1, OV9282_MODE_STREAMING);
+	ret = ov9282_write_reg(ov9282, OV9282_REG_MODE_SELECT, 1, OV9282_MODE_STREAMING);
```
Again a whitespace change, and this one breaks the linux-media coding standard of max 80 chars per line. https://www.kernel.org/doc/html/latest/driver-api/media/maintainer-entry-profile.html#coding-style-addendum
```
	ret = ov9282_write_reg(ov9282, OV9282_REG_MODE_SELECT, 1,
			       OV9282_MODE_STREAMING);
```
would be permitted and cleaner, but again it shouldn't be mixed in with a functional change.

6by9 commented on drivers/media/i2c/ov9282.c:
```diff
@@ -964,13 +1023,12 @@ static int ov9282_start_streaming(struct ov9282 *ov9282)
 		}
 	};
-	const struct ov9282_reg_list *reg_list;
+	struct ov9282_reg_list *reg_list;
```
Why have we lost the `const`?
And then we regain it in your second patch.

6by9 commented on arch/arm/boot/dts/overlays/ov9281-overlay.dts:
```diff
@@ -1,8 +1,9 @@
 /dts-v1/;
 /plugin/;
 
+#include <dt-bindings/gpio/gpio.h>
```
Why is this required? I don't see any references in the overlay to defines that it provides.

6by9 reviewed (changes requested):
Sorry but this seems to be worse than last time.

You haven't address pelwell's comment
> Please squash your ov9282 commits into one, then split that in two - one for the driver changes and one for the overlay changes. 

We have a number of whitespace changes. One is correct, but whitespace cleanups should never be merged into the same commit as functional changes.

Commit:
media: i2c: ov9282: Add external FSIN trigger snapshot mode

This patch adds support for external FSIN-triggered snapshot mode
to the OmniVision OV9282 sensor driver. It enables frame capture
synchronized with an external hardware trigger signal.

Signed-off-by: Omer Faruk Edemen <ofedemen@lectrontech.com>

Commit:
arm: dts: overlays: Add external trigger mode support

Adds DT property `trigger-mode` to enable FSIN-triggered frame capture.
Includes overlay and README update for ov9281_trig.

Signed-off-by: Omer Faruk Edemen <ofedemen@lectrontech.com>

omeredemen commented:
So i think i had made some changes in the original. and didn't squash the commits with the correct changes. i just changed the necessary part for the trigger mode and didn't make any changes on the other parts.

6by9 commented on drivers/media/i2c/ov9282.c:
```diff
@@ -998,6 +1059,16 @@ static int ov9282_start_streaming(struct ov9282 *ov9282)
+			dev_err(ov9282->dev, "failed to config external trigger mode");
+			return ret;
+		}
+		return 0;
```
There's a bit of a gotcha here with this early bail out that I hadn't noticed it in the previous review pass.

My preference would be to keep setting OV9282_REG_MODE_SELECT here in both cases, ie.
```
	/* Configure FSIN external trigger mode */
	if (ov9282->trigger_mode > 0) {
		ret = ov9282_apply_trigger_config(ov9282);
		if (ret) {
			dev_err(ov9282->dev, "failed to config external trigger mode");
			return ret;
		}
		/* stay in standby mode and wait for trigger signal */
		ret = ov9282_write_reg(ov9282, OV9282_REG_MODE_SELECT,
				       1, OV9282_MODE_STANDBY);
	} else {
		/* Start streaming */
		ret = ov9282_write_reg(ov9282, OV9282_REG_MODE_SELECT,
				       1, OV9282_MODE_STREAMING);
	}

	if (ret)
		dev_err(ov9282->dev, "fail to start streaming");

	return ret;
}
```

6by9 reviewed (commented):
One slight nasty here.

Seeing as this won't get upstreamed in this form I am prepared to accept it as is, but it would be nicer if you'd update.

Thanks.

6by9 commented on drivers/media/i2c/ov9282.c:
```diff
@@ -947,6 +956,58 @@ static int ov9282_get_selection(struct v4l2_subdev *sd,
+		return ret;
+
+	/* stay standby mode and wait for trigger signal */
+	ret = ov9282_write_reg(ov9282, OV9282_REG_MODE_SELECT,
```
Sorry I didn't say it explicitly, but we now have a duplication here. Just drop this write.
